File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)
        
        # Pair each value with its original index
        pairs = [(nums[i], i) for i in range(n)]
        
        # Sort by value
        pairs.sort(key=lambda x: x[0])
        
        # Identify groups (contiguous blocks separated by gap > limit)
        groups = []
        start = 0
        for i in range(1, n + 1):
            if i == n or pairs[i][0] - pairs[i-1][0] > limit:
                groups.append((start, i - 1))
                start = i
        
        # Process each group
        for s, e in groups:
            length = e - s + 1
            values = [pairs[i][0] for i in range(s, e + 1)]
            indices = [pairs[i][1] for i in range(s, e + 1)]
            
            # Sort indices to assign values in ascending index order
            indices.sort()
            
            for i in range(length):
                nums[indices[i]] = values[i]
        
        return nums

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<pair<int, int>> pairs; // {value, original_index}
        pairs.reserve(n);

        for (int i = 0; i < n; i++) {
            pairs.push_back({nums[i], i});
        }

        // Sort by value
        sort(pairs.begin(), pairs.end());

        // Group by limit condition
        vector<pair<int, int>> groups; // {start, end} indices in pairs array
        int start = 0;
        for (int i = 1; i <= n; i++) {
            if (i == n || pairs[i].first - pairs[i - 1].first > limit) {
                groups.push_back({start, i - 1});
                start = i;
            }
        }

        // Process each group
        for (auto [s, e] : groups) {
            int len = e - s + 1;
            vector<int> values(len);
            vector<int> indices(len);

            // Extract values and original indices
            for (int i = s, j = 0; i <= e; i++, j++) {
                values[j] = pairs[i].first;
                indices[j] = pairs[i].second;
            }

            // Sort indices to assign values in left‑to‑right order
            sort(indices.begin(), indices.end());

            for (int i = 0; i < len; i++) {
                nums[indices[i]] = values[i];
            }
        }

        return nums;
    }
};